File: enchantron-core/src/ui/sprite_source.rs
use super::{Sprite, SpriteGroup};
use crate::native::Texture;

pub trait SpriteSource: Send + Sync {
    type T: Texture;
    type S: Sprite<T = Self::T>;
    type G: SpriteGroup<T = Self::T, S = Self::S>;

    fn create_sprite(&self) -> Self::S;

    fn create_group(&self) -> Self::G;
}

File: enchantron-apple/Enchantron Shared/SpriteSource.swift
//
//  File.swift
//  Enchantron iOS
//
//  Created by Kevin Guthrie on 8/29/19.
//  Copyright © 2019 Rook And Pawn Industries, Inc. All rights reserved.
//

import SpriteKit

protocol SpriteSource {
    func createSprite() -> Sprite
    func createGroup() -> SpriteGroup
}

func createSpriteOn(parent : SpriteGroup) -> Sprite {
    let sprite = parent.createNewSprite()
    
    let onMain : () -> () = {
        parent.addSprite(sprite: sprite)
    }
    
    if Thread.isMainThread {
        onMain()
    }
    else {
        DispatchQueue.main.sync { onMain() }
    }
    
    return sprite
}

func createGroupOn(parent: SpriteGroup) -> SpriteGroup {
    let group = parent.createNewGroup()
    
    let onMain : () -> () = {
        parent.addGroup(group: group)
    }
    
    if Thread.isMainThread {
        onMain()
    }
    else {
        DispatchQueue.main.sync { onMain() }
    }
    
    return group
}

